### PR TITLE
TabPanel: add tabName prop (controlled component)

### PR DESCRIPTION
**Pull request by madhusudhand**
## What?

Add new prop `selectedTabName` to control the component tab selection programatically.

## Why?

Currently it is not possible to activate a tab using component props. This allows to activate a tab programatically.
The first usecase for this is to auto select the tab in `style book` based on the selected block in global styles.

tab-selection

### CODE REVIEW

File: packages/components/src/tab-panel/index.tsx [hunk 1]
```diff
@@ -110,6 +111,12 @@ export function TabPanel( {
 		}
 	}, [ tabs, selectedTab?.name, initialTabName, handleTabSelection ] );
 
+	useEffect( () => {
+		if ( selectedTabName ) {
+			setSelected( selectedTabName );
```
Comment (mirka) [inline packages/components/src/tab-panel/index.tsx]: I think the convention is that the `onSelect()` should only be triggered from a user interaction on a tab button, not whenever the provided value changes.

You might want to look into the [`useControlledValue` hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/hooks/use-controlled-value.ts#L21) to make the controlled/uncontrolled handling easier. The code is a bit hard to follow at the moment because of the unclear naming between `selectedTabName` and `selectedTab.name`.

File: packages/components/src/tab-panel/test/index.tsx [hunk 2]
```diff
@@ -105,6 +105,18 @@ describe( 'TabPanel', () => {
 		expect( selectedTab ).toHaveTextContent( 'Beta' );
 	} );
 
+	it( 'should render with a tab selected by prop selectedTabName', () => {
+		render(
+			<TabPanel
+				selectedTabName="beta"
+				tabs={ TABS }
+				children={ () => undefined }
+			/>
```
Comment (mirka) [inline packages/components/src/tab-panel/test/index.tsx]: I think it's worth improving the test coverage here a bit, because this isn't a sufficient representation of how controlled mode is used. I feel like we should at least have a `useState` and test a tab change, coming from both a user interaction and from a `rerender()`.

And now that we support both controlled/uncontrolled mode, I think the tests should acknowledge that. Ideally we want all the relevant functionality to be tested in both modes (abstracted into `it.each()` blocks?), but at the least the tests should be grouped into controlled/uncontrolled so that we can easily see what coverage is missing.

File: packages/components/src/tab-panel/types.ts [hunk 3]
```diff
@@ -55,6 +55,10 @@ export type TabPanelProps = {
 	 * If this prop is not set, the first tab will be selected by default.
 	 */
 	initialTabName?: string;
+	/**
+	 * The name of the tab to be selected.
+	 */
+	selectedTabName?: string;
```
Comment (mirka) [inline packages/components/src/tab-panel/types.ts]: Let's also make sure this is added to the component's readme file.

**Review by mirka — COMMENTED** [review]
Nice enhancement! I just wanted to mention as a heads-up: The disabled tab feature will be landing soon (#46471), so controlled mode also needs to take that into account.

File: packages/components/src/tab-panel/index.tsx (hunk 1 above)
Comment (madhusudhand) [reply packages/components/src/tab-panel/index.tsx]: > I think the convention is that the onSelect() should only be triggered from a user interaction on a tab button, not whenever the provided value changes

I think this is taken care. Calling `setSelected` updates the `selectedTab` object, and doesn't call `handleTabSelection` so it avoids the `onSelect` call. Can you confirm if it is still an issue.

> The code is a bit hard to follow at the moment because of the unclear naming between selectedTabName and selectedTab.name

Agree. But to maintain the consistency with the other prop `initialTabName` I had to use `selectedTabName`
I skipped updating `selectedTab` variable as it represents tab object and `selectedTabName` is a string.
Please suggest if we should rename any of them or add some comments for better readability.

File: packages/components/src/tab-panel/types.ts (hunk 3 above)
Comment (madhusudhand) [reply packages/components/src/tab-panel/types.ts]: Updated 👍

File: packages/components/src/tab-panel/test/index.tsx (hunk 2 above)
Comment (madhusudhand) [reply packages/components/src/tab-panel/test/index.tsx]: Thanks for the suggestions. I have included more tests to test the change of input values for `selectedTabName` and for `user` interactions. Also re-organized the tests for better clarity.

**Comment by madhusudhand** [comment]
> The disabled tab feature will be landing soon (https://github.com/WordPress/gutenberg/pull/46471), so controlled mode also needs to take that into account.

Thanks for the mention. Now that the PR is merged, updated the code to consider disabled tab scenario.

**Comment by ciampo** [comment]
Just linking to [this comment](https://github.com/WordPress/gutenberg/pull/46275#discussion_r1038088781) that I had made a while ago about `TabPanel` and its controlled mode. A few notes in particular:

- prop name: in my comment I proposed `tabName`, I thought it made more sense (usually inputs have a `value` prop, not a `selectedValue` prop)
- controlled/uncontrolled: have we considered using the [`useControlledValue` hook](https://github.com/WordPress/gutenberg/blob/8633ae8baf401ad3236e5c6debf0d5cb127af44a/packages/components/src/utils/hooks/use-controlled-value.ts)?
- have we checked that these API changes are still compatible with our plans of migrating this component to use `ariakit` under the hood ? (I personally don't think there should be any problems, but it would be good to double-check)
- we need to make sure that current uncontrolled usages of the component still work as expected (we should be adding specific unit tests)

File: packages/components/src/tab-panel/index.tsx (hunk 1 above)
Comment (mirka) [reply packages/components/src/tab-panel/index.tsx]: > I think this is taken care. Calling `setSelected` updates the `selectedTab` object, and doesn't call `handleTabSelection` so it avoids the `onSelect` call. Can you confirm if it is still an issue.

Yes, it looks good now 👍  I don't remember what it was before, but maybe I misread!

> Agree. But to maintain the consistency with the other prop `initialTabName` I had to use `selectedTabName` I skipped updating `selectedTab` variable as it represents tab object and `selectedTabName` is a string. Please suggest if we should rename any of them or add some comments for better readability.

One way I've seen it being handled in our codebase is to immediately rename it at destructuring time and append `Prop`. So in this case that would be to destructure our prop as `selectedTabName: selectedTabNameProp`. (Search the codebase for `valueProp` for some examples of this pattern to distinguish a prop value from an internal state value.)

File: packages/components/src/tab-panel/test/index.tsx [hunk 4]
```diff
@@ -93,193 +93,298 @@ describe( 'TabPanel', () => {
...
+					initialTabName="alpha"
+					tabs={ TABS }
+					children={ () => undefined }
+				/>
+			);
+			expect( getSelectedTab() ).toHaveTextContent( 'Beta' );
+		} );
 
-		await user.click( screen.getByRole( 'tab', { name: 'Beta' } ) );
+		it( 'should fallback to initial tab if active tab is removed', async () => {
```
Comment (madhusudhand) [inline packages/components/src/tab-panel/test/index.tsx]: with the current change, this test case fails as it fallbacks to `first active tab` irrespective of `initialTabName` is set or not, when the active tab is removed. I think it should be the right behaviour.

File: packages/components/src/tab-panel/test/index.tsx [hunk 5]
```diff
@@ -93,193 +93,298 @@ describe( 'TabPanel', () => {
...
-		);
+			rerender(
+				<TabPanel
+					tabs={ TABS.slice( 2 ) /* remove alpha, beta */ }
+					children={ () => undefined }
+					onSelect={ mockOnSelect }
+				/>
+			);
+			expect( getSelectedTab() ).toHaveTextContent( 'Gamma' );
+			expect( mockOnSelect ).toHaveBeenLastCalledWith( 'gamma' );
```
Comment (madhusudhand) [inline packages/components/src/tab-panel/test/index.tsx:262]: `onSelect` isn't called in this scenario because it is a `non-user` interaction. but it doesn't feel right.

File: packages/components/src/tab-panel/index.tsx [hunk 6]
```diff
@@ -100,28 +102,12 @@ export function TabPanel( {
...
-	}, [
-		tabs,
-		selectedTab?.name,
-		selectedTab?.disabled,
-		initialTabName,
-		handleTabSelection,
-	] );
+	// useEffect( () => {
+	// 	setTabName( selectedTab?.name );
+	// }, [ selectedTab?.name ] );
```
Comment (madhusudhand) [inline packages/components/src/tab-panel/index.tsx]: Without this hook, `onSelect` is not called on non user interactions such as changes to prop `tabName`, or removal of selected tab from `tabs`. Having this hook creates a side effect of calling the `onSelect` method twice for user actions such as click.

Please suggest ideas to handle this.

**Comment by ciampo** [comment]
This PR needs a rebase after the fix in #47100 was merged. I can take another look at the code once the rebase has happened :)

File: packages/components/src/tab-panel/test/index.tsx (hunk 4 above)
Comment (ciampo) [reply packages/components/src/tab-panel/test/index.tsx]: Let's see if there's any change here after rebasing and including changes from  https://github.com/WordPress/gutenberg/pull/47100

**Comment by ciampo** [comment]
Hey @madhusudhand ,  are you still able to work on this PR ?

**Comment by madhusudhand** [comment]
Hey @ciampo 

Updated the component and following are the test scenarios.

image

File: packages/components/src/tab-panel/test/index.tsx [hunk 7]
```diff
@@ -9,7 +9,12 @@ import userEvent from '@testing-library/user-event';
  */
 import TabPanel from '..';
 
-const setupUser = () => userEvent.setup();
+jest.useFakeTimers();
```
Comment (ciampo) [inline packages/components/src/tab-panel/test/index.tsx]: We've been recently working across the codebase to refactor unit tests _away_ from using take timers and `act()` call.

That would mean removing the `jest.useFakeTimers();` call and the `advanceTimers: jest.advanceTimersByTime,` option when setting `userEvent` up.

cc @tyxla

File: packages/components/src/tab-panel/test/index.tsx [hunk 8]
```diff
@@ -49,289 +54,404 @@ describe( 'TabPanel', () => {
...
-		await user.click( screen.getByRole( 'tab', { name: 'Alpha' } ) );
-
-		expect( getSelectedTab() ).toHaveTextContent( 'Alpha' );
-		expect(
-			screen.getByRole( 'tabpanel', { name: 'Alpha' } )
-		).toBeInTheDocument();
-		expect( panelRenderFunction ).toHaveBeenLastCalledWith( TABS[ 0 ] );
-		expect( mockOnSelect ).toHaveBeenLastCalledWith( 'alpha' );
-	} );
+	describe( 'Uncontrolled mode', () => {
```
Comment (ciampo) [inline packages/components/src/tab-panel/test/index.tsx]: A pattern that we've started to adopt when writing unit tests for components that can be used both in controlled and uncontrolled mode, is to use `describe.each` and run the same suite against the controlled and uncontrolled version.

You can see it in action for the `ComboboxControl` component:

https://github.com/WordPress/gutenberg/blob/56c33cf1f529d286b9c0db49534a2f5617fdcc84/packages/components/src/combobox-control/test/index.js#L80-L83


Of course, if there are certain tests that are specifically aimed at one version of the component, they can be added outside of the `describe.each` call.

**Review by ciampo — COMMENTED** [review]
Thank you @madhusudhand for your work here!

It actually looks like there's quite a few scenarios that we'd need to test, especially around:
- the initial tab selection 
- the currently selected tab becoming disabled
- the component in general working as expected in controlled and uncontrolled mode alike

Since you've added also a bunch of unit tests (which is great!), I suggest we do the following:

1. We open a new PR, where we add the new unit tests to the current implementation of `TabPanel`. Of course, we'll be able only to test the component in uncontrolled mode, but adding those tests separately can help us to certify the behaviour of the component on `trunk`
2. Once the PR with the unit tests is merged, we can rebase this PR and just make the necessary changes to include the new `tabName` prop and enable controlled mode. We will just need to run the existing tests on the controlled version of the component, and make adjustments if necessary.

I believe that the suggested approach will enable us to move faster and will give us more confidence when merging such changes.

What do folks think?

File: packages/components/src/tab-panel/test/index.tsx (hunk 7 above)
Comment (tyxla) [reply packages/components/src/tab-panel/test/index.tsx]: Yes, we use fake timers only when we really have to. That means removing `jest.useFakeTimers()` and calling `const user = userEvent.setup();` without the `advanceTimers` setting.

Here are some PRs where we refactored from fake to real timers if that will help:

* https://github.com/WordPress/gutenberg/pull/47318
* https://github.com/WordPress/gutenberg/pull/47218
* https://github.com/WordPress/gutenberg/pull/47144
* https://github.com/WordPress/gutenberg/pull/47056
* https://github.com/WordPress/gutenberg/pull/46714

**Comment by ciampo** [comment]
Hey @madhusudhand , do you still have the capacity to work on this?

**Comment by madhusudhand** [comment]
@ciampo I would be able to continue the work after 16th Feb. If you have capacity feel free to take it from me. Thanks.

**Comment by aaronrobertshaw** [comment]
> We open a new PR, where we add the new unit tests to the current implementation of TabPanel. Of course, we'll be able only to test the component in uncontrolled mode, but adding those tests separately can help us to certify the behaviour of the component on trunk

I put together a draft PR (https://github.com/WordPress/gutenberg/pull/48086) that you might use as a base for verifying the current component behaviour on trunk. I'll be AFK for the next week so feel free to pick it up or close it if it doesn't help.

**Comment by madhusudhand** [comment]
**Update**: Rebased the PR with unit testes from #48086 and they all are running great. Will push more tests for controlled mode.

**Comment by ciampo** [comment]
> **Update**: Rebased the PR with unit testes from #48086 and they all are running great. Will push more tests for controlled mode.

Nice! You can take inspiration from `ComboboxControl`'s tests on how to add the controlled version of the component to the tests:

https://github.com/WordPress/gutenberg/blob/dc32380b2dc744a8cfb7725c0d7f145a471ec3f3/packages/components/src/combobox-control/test/index.tsx#L61-L85

**Comment by madhusudhand** [comment]
Updated the PR with revised changes and included unit tests for controlled and uncontrolled behaviours.

Note: removed `useControlledValue` hook from this revision as it is becoming redundant with `Ariakit.useTabStore`

**Comment by ciampo** [comment]
Hey @madhusudhand , thank you for continuing the work here!

In the past months, we've started working on improving the component:

- We've [refactored `TabPanel` to Ariakit](https://github.com/WordPress/gutenberg/pull/52133)
- We started work on [a new `Tabs` component](https://github.com/WordPress/gutenberg/pull/53960) — the name better follows the WAI-ARIA pattern, and having a new component allows us to introduce better APIs (including controlled mode and granular components for better composition)

The plan is to leave `TabPanel` around for backwards compatibility reasons, but to stop maintaining it (if not for bug fixes).

Furthermore, these components have quite an opinionated behaviour when uncontrolled around making sure that a tab gets selected even when the currently selected tab becomes disabled / gets removed. When allowing the component to be controlled, we have to ask ourselves how much of that custom behaviour we want to inherit (you can follow the conversation in https://github.com/WordPress/gutenberg/pull/53960 ).

**Comment by madhusudhand** [comment]
Closing this change. Use the [new Tabs component](https://github.com/WordPress/gutenberg/pull/53960) instead.
